File: my app/src/stt.js
const childProcess = require("node:child_process");
const fs = require("node:fs");
const path = require("node:path");

function ffmpegAvailable() {
  try {
    const r = childProcess.spawnSync("ffmpeg", ["-version"], { stdio: "ignore" });
    return r.status === 0;
  } catch {
    return false;
  }
}

async function downloadTelegramFile(ctx, fileId, outPath) {
  const url = await ctx.telegram.getFileLink(fileId);
  const res = await fetch(url.href);
  if (!res.ok) throw new Error(`Download failed: ${res.status}`);
  const buf = Buffer.from(await res.arrayBuffer());
  fs.writeFileSync(outPath, buf);
  return outPath;
}

function convertToWav(inPath, outPath) {
  const r = childProcess.spawnSync(
    "ffmpeg",
    ["-y", "-i", inPath, "-ac", "1", "-ar", "16000", outPath],
    { stdio: "ignore" }
  );
  return r.status === 0 && fs.existsSync(outPath);
}

async function transcribeViaOpenAI(wavPath) {
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) return null;

  // Minimal HTTP call. If you want, we can swap to the official OpenAI Node SDK.
  const form = new FormData();
  form.append("model", process.env.OPENAI_STT_MODEL || "whisper-1");
  form.append("file", new Blob([fs.readFileSync(wavPath)]), path.basename(wavPath));

  const res = await fetch("https://api.openai.com/v1/audio/transcriptions", {
    method: "POST",
    headers: { Authorization: `Bearer ${apiKey}` },
    body: form,
  });

  if (!res.ok) return null;
  const data = await res.json().catch(() => null);
  return typeof data?.text === "string" ? data.text : null;
}

async function transcribeVoice(ctx, fileId, tmpDir) {
  const backend = String(process.env.STT_BACKEND || "none").trim().toLowerCase();
  if (!backend || backend === "none" || backend === "off" || backend === "false" || backend === "0") {
    await ctx.reply(
      "Ovozdan qidirish hozir o‘chiq.\n" +
        "Agar xohlasangiz `.env` da `STT_BACKEND=openai` qiling (va `OPENAI_API_KEY` qo‘ying)."
    );
    return null;
  }

  if (backend !== "openai") {
    await ctx.reply(`STT_BACKEND noto‘g‘ri: <code>${backend}</code>`, { parse_mode: "HTML" });
    return null;
  }

  const oggPath = path.join(tmpDir, "voice.ogg");
  const wavPath = path.join(tmpDir, "voice.wav");

  await downloadTelegramFile(ctx, fileId, oggPath);

  if (ffmpegAvailable()) {
    const ok = convertToWav(oggPath, wavPath);
    if (!ok) {
      await ctx.reply("Ovozni konvertatsiya qila olmadim (ffmpeg xatosi).");
      return null;
    }
  } else {
    // OpenAI odatda ko‘p formatlarni qabul qiladi, lekin konvert bo‘lmasa baribir urinamiz.
    fs.copyFileSync(oggPath, wavPath);
  }

  const text = await transcribeViaOpenAI(wavPath);
  if (text == null) {
    await ctx.reply(
      "Ovozdan matn olish ishlamadi.\n" +
        "Tekshiring: `OPENAI_API_KEY` to‘g‘ri va hisobingizda balans bor.\n" +
        "Agar xohlasangiz, lokal (offline) STT variantini ham qo‘shib beraman."
    );
    return null;
  }
  return text;
}

module.exports = { transcribeVoice };

File: my app/stt.py
from __future__ import annotations

import os
import shutil
import subprocess
import tempfile
from typing import Optional

from aiogram import Bot
from aiogram.types import Message


def _ffmpeg_available() -> bool:
    return shutil.which("ffmpeg") is not None


async def transcribe_voice_message(message: Message, bot: Bot) -> Optional[str]:
    """
    Telegram voice message (.ogg/.opus) -> matn.
    STT_BACKEND:
      - none: o‘chirib qo‘yilgan
      - faster-whisper: lokal Whisper (faster-whisper) orqali
    """
    backend = (os.getenv("STT_BACKEND") or "none").strip().lower()
    if backend in {"", "none", "off", "false", "0"}:
        await message.answer(
            "Ovozdan qidirish hozir o‘chiq.\n"
            "Agar xohlasangiz `.env` da `STT_BACKEND=faster-whisper` qiling."
        )
        return None

    if backend != "faster-whisper":
        await message.answer(f"STT_BACKEND noto‘g‘ri: <code>{backend}</code>")
        return None

    if not message.voice:
        return None

    if not _ffmpeg_available():
        await message.answer(
            "Ovozli xabarni o‘qish uchun `ffmpeg` kerak (PATH’da bo‘lsin).\n"
            "FFmpeg o‘rnatilgandan keyin qayta urinib ko‘ring."
        )
        return None

    file = await bot.get_file(message.voice.file_id)
    voice_bytes = await bot.download_file(file.file_path)

    with tempfile.TemporaryDirectory() as td:
        ogg_path = os.path.join(td, "voice.ogg")
        wav_path = os.path.join(td, "voice.wav")
        with open(ogg_path, "wb") as f:
            f.write(voice_bytes.read())

        # 16kHz mono wav
        proc = subprocess.run(
            ["ffmpeg", "-y", "-i", ogg_path, "-ac", "1", "-ar", "16000", wav_path],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            check=False,
        )
        if proc.returncode != 0 or not os.path.exists(wav_path):
            await message.answer("Ovozni konvertatsiya qila olmadim (ffmpeg xatosi).")
            return None

        try:
            from faster_whisper import WhisperModel
        except Exception:
            await message.answer(
                "`faster-whisper` o‘rnatilmagan yoki import bo‘lmadi.\n"
                "O‘rnatish: <code>python -m pip install -r requirements.txt</code>"
            )
            return None

        model_name = (os.getenv("WHISPER_MODEL") or "base").strip()
        model = WhisperModel(model_name, device="cpu", compute_type="int8")
        segments, _info = model.transcribe(wav_path, beam_size=5, vad_filter=True)
        text = "".join(seg.text for seg in segments).strip()
        return text

